src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs

```diff
@@ -487,54 +475,94 @@ private void UpdateChangeSet(DbContext context, EntityChangeSet changeSet)
                 var entityType = GetEntityType(context.As<IObjectContextAdapter>().ObjectContext, entityEntry.GetEntityBaseType(), useClrType: false);
                 entityChange.EntityId = GetEntityId(entityEntry, entityType);
 
-                /* Update foreign keys */
+                /* Add/Remove/Update foreign keys*/
+                var foreignKeyPropertyNames = entityType.NavigationProperties
+                                                .SelectMany(foreignKeys => foreignKeys.GetDependentProperties())
+                                                .Select(p => p.Name)
+                                                .Distinct()
+                                                .ToList();
 
-                var foreignKeys = entityType.NavigationProperties;
-
-                foreach (var foreignKey in foreignKeys)
+                foreach(var propertyName in foreignKeyPropertyNames)
                 {
-                    foreach (var property in foreignKey.GetDependentProperties())
+                    var propertyEntry = entityEntry.Property(propertyName);
+                    var propertyInfo = entityEntry.GetPropertyInfo(propertyName);
+                    var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == propertyName);
+                    var oldValue = propertyEntry.GetOriginalValue();
+                    var newValue = propertyEntry.GetNewValue();
+
+                    if (propertyChange == null)
                     {
-                        var propertyEntry = entityEntry.Property(property.Name);
-                        var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == property.Name);
+                        if (propertyEntry.HasChanged())
+                        {
+                            // Add foreign key
+                            entityChange.PropertyChanges.Add(CreateEntityPropertyChange(oldValue, newValue, propertyInfo));
+                        }
+
+                        continue;
+                    }
 
-                        //make sure test case cover post saving (for foreign key update)
-                        if (propertyChange == null)
+                    if (propertyChange.OriginalValue == propertyChange.NewValue)
+                    {
+                        var newPropertyChange = CreateEntityPropertyChange(oldValue, newValue, propertyInfo);
+                        if (propertyChange.NewValue == newPropertyChange.NewValue)
```

Comment:
The instance was created because of the `Truncation` in the new constructor. 🙈

Reply:
Yes, I noticed. But that means caller must know implementation detail of the constructor.
So I suggested to expose an explicit setter (`SetNewValue`) and checker (`NewValueEquals`).

Actually, it should be the responsibility of the creator (`EntityHistoryHelper`) to handle such logic outside the entity (since the max length is public information), as it was before this PR.

Reply:
> Actually, it should be the responsibility of the creator (EntityHistoryHelper) to handle such logic outside the entity (since the max length is public information), as it was before this PR.

Agreed. I was trying to follow the similar approach of `SetNormalizedNames()` on `AbpUser` and `AbpRole`. I guess they aren't the same use case then.

I am thinking of creating a private variable to store the original `NewValue`.
So it will work like this:
- `SetNewValue()` will store the original value to the private variable before truncating the value to be stored in `NewValue`
- `NewValueEquals()` will use private variable to perform comparison.
- protected setter for `NewValue` will call `SetNewValue()`

Reply:
> Agreed. I was trying to follow the similar approach of `SetNormalizedNames()` on `AbpUser` and `AbpRole`. I guess they aren't the same use case then.

Yes, they aren't the same.

> - protected setter for `NewValue` will call `SetNewValue(string)`

That's not possible due to recursion (you can try).

Reply:
In that case, how about keep the setter public and implements the truncation and store it into private variable in the setter

Reply:
I will leave this to a separate PR